Add rules for `ref` safety

BillWagner opened the pull request:
This updates the **Variables** clause with several sections for `ref` safety.

This PR is primarily from taking the text from the [Span safety](https://learn.microsoft.com/dotnet/csharp/language-reference/proposals/csharp-7.2/span-safety) proposal and editing into the style for the standard.

In addition, a few rules were added:

1. `out` parameters can't be returned by `ref`
2. An instance method of a `struct` can't return `this` by `ref`.
3. An `in` parameter can't be returned by `ref`, only by `ref readonly`. 

The third rules makes this illegal:

```csharp
public ref int MakeWritable(in int p)
{
    return ref p; // Error,
 }

readonly int j = 5;

ref int k = MakeWritable(in j);
k = 7;
```

BillWagner committed:
first pass at safety rules

This mostly incorporates the feature spec language.

BillWagner committed:
Add rule for `out` parameters

Out parameters have a ref safe to escape scope of the current method, not the calling method.

BillWagner committed:
Add readonly rule.

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,139 @@ variable_reference
+
+### §ref-span-safety-general General
+
+A *reference variable* is a local variable or `struct` declared with the `ref` modifier. A ref local does not create a new storage location. Instead, a ref ref local represents the same storage location as its initializing expression. Thus, the value of a reference variable is always the same as the underlying variable.
```
I'm confused as to what a local variable and a type have in common - it would be odd to call the struct itself a variable.

Additionally, is "ref ref local" meant to be "ref local"?

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,139 @@ variable_reference
+
+A *reference variable* is a local variable or `struct` declared with the `ref` modifier. A ref local does not create a new storage location. Instead, a ref ref local represents the same storage location as its initializing expression. Thus, the value of a reference variable is always the same as the underlying variable.
+
+A `ref struct` is a `struct` whose fields include a `ref` struct or a ref-like field. A ref-like field does not create a new storage location. Instead, if refers to the same storage as its initializing expression. Ref struct types include `Span<T>`, `ReadOnlySpan<T>`, and other types that include an unmanaged pointer as a member.
```
I'd expect a `ref struct` to be ` struct` declared with the `ref` modifier - with restrictions on when that modifier can be applied. (I'd expect it to be *possible* to declare a `ref struct` without any fields that are ref structs or ref-like fields, even if it would normally be pointless to do so.)

Additionally, should "`ref` struct" be "`ref struct`"? (Or just ref struct? Currently we've got three variations of using code font in this paragraph...)

BillWagner replied:
I made updates. On the first point, what's important here is that `ref struct` types have "ref-like" fields, so therefore have specific rules in this section.

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,139 @@ variable_reference
+
+### §ref-span-safety-escape-scopes Safe to escape scopes
+
+At compile-time, each expression is associated with a scope that expression is permitted to escape to, *safe-to-escape-scope*. Each variable is associated with the scope a reference to it is permitted to escape to, *ref-safe-to-escape-scope*. For a given variable expression, these may be different. When the *ref-safe-to-escape* scope is *calling-method*, the variable is *safe-to-return*. A variab ...
```
```suggestion
At compile-time, each expression is associated with a scope that expression is permitted to escape to, its *safe-to-escape-scope*. Each variable is associated with the scope a reference to it is permitted to escape to, *ref-safe-to-escape-scope*. For a given variable expression, these may be different. When the *ref-safe-to-escape* scope is *calling-method*, the variable is *safe-to-return*. A variable that is *safe-to-return* may escape the enclosing method as a whole. In other words, the variable can *returned-by-ref*.
```

Or something similar - basically having *safe-to-escape-scope* on its own after the comma reads a bit oddly to me. Ditto for the second sentence. Another option would be to move the name earlier, e.g.

> At compile-time, each expression is associated with a *safe-to-escape-scope*, which is the scope that the expression is permitted to escape to.

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,139 @@ variable_reference
+
+At compile-time, each expression is associated with a scope that expression is permitted to escape to, *safe-to-escape-scope*. Each variable is associated with the scope a reference to it is permitted to escape to, *ref-safe-to-escape-scope*. For a given variable expression, these may be different. When the *ref-safe-to-escape* scope is *calling-method*, the variable is *safe-to-return*. A variab ...
+
+Given an assignment from an expression E1 with a *safe-to-escape-scope* S1, to a (variable) expression E2 with *safe-to-escape-scope* S2, it is an error if S2 is a wider scope than S1. By construction, the two scopes S1 and S2 are in a nesting relationship, because a legal expression is always *safe-to-return* from some scope enclosing the expression.
```
Do we define "wider" anywhere for scopes? (Can scopes ever overlap without nesting?)

BillWagner replied:
Let's discuss the terms in our meeting Wednesday.  The standard defines "scope" in S 7

> The scope of a name is the region of program text within which it is possible to refer to the entity declared by the name without qualification of the name. 

In this section, we're using "scope" to address a related to different concept: The region where the storage referred to by an "initializing expression" is valid. Maybe a different term is better? 

However, to answer your question, as defined in "Basic Concepts", scopes can be nested, but don't overlap.

BillWagner replied:
I updated the definitions in the general section for reference variables.

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,139 @@ variable_reference
+
+Given an assignment from an expression E1 with a *safe-to-escape-scope* S1, to a (variable) expression E2 with *safe-to-escape-scope* S2, it is an error if S2 is a wider scope than S1. By construction, the two scopes S1 and S2 are in a nesting relationship, because a legal expression is always *safe-to-return* from some scope enclosing the expression.
+
+There are three different values for *ref-safe-to-escape-scope*: *block in which variable is declared*, *current-method* and *calling-method*. When the *ref-safe-to-escape-scope* is *calling-method*, the variable is *safe-to-return* by reference.
```
It feels odd to me that one of these names is space-separated, and the others are dash-separated.

BillWagner replied:
Fixed. I'd really like to find a shorter term for *block-in-which-variable-is-declared* though.

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,139 @@ variable_reference
+
+### §ref-span-safety-locals Local variable escape scope
+
+For a local variable `l`:
```
I'd suggest using a different name here, for all the normal reasons that `l` is often hard to read or ambiguous. Perhaps `v`?

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,139 @@ variable_reference
+
+Given an assignment from an expression E1 with a *safe-to-escape-scope* S1, to a (variable) expression E2 with *safe-to-escape-scope* S2, it is an error if S2 is a wider scope than S1. By construction, the two scopes S1 and S2 are in a nesting relationship, because a legal expression is always *safe-to-return* from some scope enclosing the expression.
+
+There are three different values for *ref-safe-to-escape-scope*: *block in which variable is declared*, *current-method* and *calling-method*. When the *ref-safe-to-escape-scope* is *calling-method*, the variable is *safe-to-return* by reference.
```
Also, should calling-method actually be caller-scope or something like that? (Just thinking that the caller might not be a method...)

BillWagner replied:
I like caller-scope. Agreed that it could be from a property or index accessor.

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,139 @@ variable_reference
+
+- A local whose type is not a `ref struct` type is *safe-to-return* from the entire enclosing method. Its *safe-to-escape-scope* is the *calling-method*.
+- If the variable is an iteration variable of a `foreach` loop, then the variable's *safe-to-escape-scope* is the same as the *safe-to-escape-scope* of the `foreach` loop's expression.
+- A local of `ref struct` type and uninitialized at the point of declaration is *safe-to-return* from the entire enclosing method. Its *safe-to-escape-scope* is the calling method.
```
```suggestion
- A local of `ref struct` type which is uninitialized at the point of declaration is *safe-to-return* from the entire enclosing method. Its *safe-to-escape-scope* is the *calling method*.
```

(Or possible *calling-method*. We need to work out how we want to deal with these, then be consistent.)

BillWagner committed:
fix headers

no code fences in headers

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,139 @@ variable_reference
+- A local whose type is not a `ref struct` type is *safe-to-return* from the entire enclosing method. Its *safe-to-escape-scope* is the *calling-method*.
+- If the variable is an iteration variable of a `foreach` loop, then the variable's *safe-to-escape-scope* is the same as the *safe-to-escape-scope* of the `foreach` loop's expression.
+- A local of `ref struct` type and uninitialized at the point of declaration is *safe-to-return* from the entire enclosing method. Its *safe-to-escape-scope* is the calling method.
+- Otherwise the variable's type is a `ref struct` type, and the variable's declaration requires an initializer. The variable's *safe-to-escape-scope* is the same as the *safe-to-escape* of its initializer.
```
Requires an initializer, or has an initializer?

BillWagner replied:
has is probably better.

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,139 @@ variable_reference
+
+The application of a user-defined operator is treated as a method invocation.
+
+For an operator that yields an value, such as `e1 + e2` or `c ? e1 : e2`, the *safe-to-escape-scope* of the result is the narrowest scope among the *safe-to-escape-scopes* of the operands of the operator.  As a consequence, for a unary operator that yields an value, such as `+e`, the *safe-to-escape-scope* of the result is the *safe-to-escape-scope* of the operand.
```
```suggestion
For an operator that yields a value, such as `e1 + e2` or `c ? e1 : e2`, the *safe-to-escape-scope* of the result is the narrowest scope among the *safe-to-escape-scopes* of the operands of the operator.  As a consequence, for a unary operator that yields an value, such as `+e`, the *safe-to-escape-scope* of the result is the *safe-to-escape-scope* of the operand.
```

And do we formally define "narrowest scope" anywhere? (We may decide we don't need to.)

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,139 @@ variable_reference
+
+For an operator that yields an value, such as `e1 + e2` or `c ? e1 : e2`, the *safe-to-escape-scope* of the result is the narrowest scope among the *safe-to-escape-scopes* of the operands of the operator.  As a consequence, for a unary operator that yields an value, such as `+e`, the *safe-to-escape-scope* of the result is the *safe-to-escape-scope* of the operand.
+
+For an operator that yields an variable, such as `c ? ref e1 : ref e2`:
```
```suggestion
For an operator that yields a variable, such as `c ? ref e1 : ref e2`:
```

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,139 @@ variable_reference
+
+### §ref-span-safety-method-invocation Method invocation
+
+For a variable `c` resulting from a ref-returning method invocation, `ref e1.M(e2, ...)`, its *ref-safe-to-escape-scope* is the smallest of the following scopes:
```
Narrowest instead of smallest, for consistency?

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,139 @@ variable_reference
+>
+> *end example*
+
+For a value resulting from a method invocation `e1.M(e2, ...)`, its *safe-to-escape-scope* from the smallest of the following scopes:
```
Narrowest?

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,139 @@ variable_reference
+- The *calling-method*
+- The *safe-to-escape* of all argument expressions (including the receiver)
+
+A property invocation (either `get` or `set`) it treated as a method invocation of the underlying method by the above rules.
```
```suggestion
A property invocation (either `get` or `set`) is treated as a method invocation of the underlying method by the above rules.
```

BillWagner commented on standard/variables.md:
```diff
@@ -901,3 +901,139 @@ variable_reference
+- A local of `ref struct` type and uninitialized at the point of declaration is *safe-to-return* from the entire enclosing method. Its *safe-to-escape-scope* is the calling method.
+- Otherwise the variable's type is a `ref struct` type, and the variable's declaration requires an initializer. The variable's *safe-to-escape-scope* is the same as the *safe-to-escape* of its initializer.
+
+### §ref-span-safety-parameters Parameter escape scope
```
This section doesn't handle a rule needed later in C# 11 (when `ref` fields are allowed):

-  A `ref` parameter can't be returned by `ref` as a field in another `ref` argument. 

For example:

```csharp
ref struct S { 
  ref int refField;
}

void M(ref S p1, ref int p2) {
   p1.refField = ref p2;
}
```

These rules would allow that because both `p1` and `p2` have *ref-safe-to-escape-scope* of *calling-method*. But, doing so could be dangerous:  `p2` could be a local in the calling method, and `p1` could be a static field in the type of the calling method.

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,139 @@ variable_reference
+
+### §ref-span-safety-stackalloc `stackalloc`
+
+For a value `c` resulting from a `stackalloc` expression its *safe-to-escape-scope* to the current method.
```
```suggestion
The *safe-to-escape-scope* for a value `c` resulting from a `stackalloc` expression is the current method.
```

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,139 @@ variable_reference
+
+### §ref-span-safety-default-expressions `default` expressions
+
+For a `default` expression, the *safe-to-escape-scope* is the *calling-method*.
```
```suggestion
The *safe-to-escape-scope* of a `default` expression is the *calling-method*.
```

(This and the previous suggestion just feel like more straightforward statements.)

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,139 @@ variable_reference
+
+### §ref-span-safety-limitations Limitations on reference variables
+
+- Neither a `ref` parameter, nor a `ref` local, nor a parameter or local of a `ref struct` type can be lifted into a lambda or local function.
```
```suggestion
- Neither a `ref` parameter, nor a `ref` local, nor a parameter or local of a `ref struct` type can be lifted into a lambda expression or local function.
```

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,139 @@ variable_reference
+### §ref-span-safety-limitations Limitations on reference variables
+
+- Neither a `ref` parameter, nor a `ref` local, nor a parameter or local of a `ref struct` type can be lifted into a lambda or local function.
+- Neither a `ref` parameter nor a parameter of a `ref struct` type may be an argument on an iterator method or an `async` method.
```
```suggestion
- Neither a `ref` parameter nor a parameter of a `ref struct` type may be an argument for an iterator method or an `async` method.
```

(Or perhaps "provided to"?)

jskeet reviewed (commented):
This is a very shallow review - I'll do a more complete one when I have more mental energy. (I get very confused by all of this.)

cston commented on standard/variables.md:
```diff
@@ -901,3 +901,139 @@ variable_reference
+
+### §ref-span-safety-escape-scopes Safe to escape scopes
+
+At compile-time, each expression is associated with a scope that expression is permitted to escape to, *safe-to-escape-scope*. Each variable is associated with the scope a reference to it is permitted to escape to, *ref-safe-to-escape-scope*. For a given variable expression, these may be different. When the *ref-safe-to-escape* scope is *calling-method*, the variable is *safe-to-return*. A variab ...
```
The first sentence refers to "each expression", the second refers to "each variable", and the third refers to "a given variable expression". Are the sentences referring to the same items? If so, consider using a common name - perhaps "expression", or if this is specifically about `ref` cases, perhaps "ref expression". (Are the expressions always variables or is the expression in `return ref F();` also an example?)

> When the *ref-safe-to-escape* scope is *calling-method*, the variable is *safe-to-return*. A variable that is *safe-to-return* may escape the enclosing method as a whole. In other words, the variable can *returned-by-ref*.

When the ref-safe-to-escape scope is _calling-method_, the variable is safe-to-return _by-ref_. (When the safe-to-escape scope is _calling-method_, the variable is safe-to-return _by-value_.)

BillWagner replied:
I need to think about this one more. This was intentional, but it's awkward. Not all expressions can be ref assigned or returned. Only variables (which have defined storage) can be assigned to a ref vaiable.

I'll keep thinking on the wording here.

cston commented on standard/variables.md:
```diff
@@ -901,3 +901,139 @@ variable_reference
+
+At compile-time, each expression is associated with a scope that expression is permitted to escape to, *safe-to-escape-scope*. Each variable is associated with the scope a reference to it is permitted to escape to, *ref-safe-to-escape-scope*. For a given variable expression, these may be different. When the *ref-safe-to-escape* scope is *calling-method*, the variable is *safe-to-return*. A variab ...
+
+Given an assignment from an expression E1 with a *safe-to-escape-scope* S1, to a (variable) expression E2 with *safe-to-escape-scope* S2, it is an error if S2 is a wider scope than S1. By construction, the two scopes S1 and S2 are in a nesting relationship, because a legal expression is always *safe-to-return* from some scope enclosing the expression.
```
> ... because a legal expression is always *safe-to-return* from some scope enclosing the expression.

I may be misunderstanding, but a `ref struct` value is not necessarily *safe-to-return* from the enclosing scope. For instance, the expression in `return s;` is not *safe-to-return*:
```csharp
static Span<int> F()
{
    Span<int> s = stackalloc int[1];
    return s; // error
}
```

BillWagner replied:
I think that clause can be deleted. It doesn't add anything.

BillWagner replied:
This is where I've added a sentence defining "wider" and "narrower" scopes.

cston commented on standard/variables.md:
```diff
@@ -901,3 +901,139 @@ variable_reference
+For a variable designating a reference to a field, `e.F`:
+
+- If `e` is of a reference type, its *ref-safe-to-escape-scope* is the *calling-method*.
+- Otherwise, if `e` is of a value type, its *ref-safe-to-escape-scope* is taken from the *ref-safe-to-escape* of `e`.
```
Should this section also include ref fields? If the field `e.F` is a ref field, where `e` is a of `ref struct` type, the _ref-safe-to-escape-scope_ of `ref e.F` is the _safe-to-escape_ of `e`.

BillWagner replied:
Not yet. The committee is on V7.x, so ref fields aren't supported yet.

BillWagner committed:
Apply suggestions from code review

Co-authored-by: Jon Skeet <skeet@pobox.com>

BillWagner committed:
updates from reviews

Updates from code review

BillWagner committed:
style fix

RikkiGibson commented on standard/variables.md:
```diff
@@ -901,3 +901,139 @@ variable_reference
+
+A *reference variable* is a local variable declared with the `ref` modifier, or an instance of a `ref struct` type. A ref local does not create a new storage location. Instead, a ref local represents the same storage location as its initializing expression. Thus, the value of a reference variable is always the same as the underlying variable.
+
+A `ref struct` may include `ref struct` or ref-like fields. A ref-like field does not create a new storage location. Instead, if refers to the same storage as its initializing expression. Ref struct types include `Span<T>`, `ReadOnlySpan<T>`, and other types that include an unmanaged pointer as a member.
```
> A ref-like field does not create a new storage location

This doesn't feel right to me. A ref field, or a field of ref struct type, is a storage location, and can be assigned using `v.refField = ref expr` or `v.refStructField = expr` respectively. In the ref field case, it just happens that the thing the storage location contains is a ref.

BillWagner replied:
Good catch. I did rework these paragraphs. still not sure it's done yet.

BillWagner committed:
respond to feedback

Ref like fields do have storage, but that storage may refer to a variable that is a struct parameter or varialbe.

Nigel-Ecma commented on standard/variables.md:
```diff
@@ -901,3 +901,141 @@ variable_reference
+
+### §ref-span-safety-general General
+
+A *reference variable* is a local variable declared with the `ref` modifier, or an instance of a `ref struct` type. A ref local does not create a new storage location. Instead, a ref local represents the same storage location as its initializing expression. Thus, the value of a reference variable is always the same as the underlying variable.
```
A, say, `int` local variable declaration creates a storage location of the bits of an `int` value; if  a `ref int` is declared where are the bits of the `ref int` value stored? C++ says “*It is unspecified whether or not a reference requires storage*”, are we saying in C# a reference variable *never* requires storage?

Nigel-Ecma commented on standard/variables.md:
```diff
@@ -901,3 +901,141 @@ variable_reference
+
+### §ref-span-safety-escape-scopes Safe to escape scopes
+
+At compile-time, each expression is associated with a scope that expression is permitted to escape to, its *safe-to-escape-scope*. Each variable is associated with the scope a reference to it is permitted to escape to, *ref-safe-to-escape-scope*. For a given variable expression, these may be different. When the *ref-safe-to-escape-scope* scope is *caller-scope*, the variable is *safe-to-return*.  ...
```
Does the *expression* escape or does it’s value/result escape?

BillWagner replied:
It should be variable (for ref safe to escape) or value (for safe to escape). Values only matter if they are of a type that has a ref-like field.

Nigel-Ecma commented on standard/variables.md:
```diff
@@ -901,3 +901,141 @@ variable_reference
+
+There are three different values for *ref-safe-to-escape-scope*: *block-in-which-variable-is-declared*, *current-method* and *caller-scope*. When the *ref-safe-to-escape-scope* is *caller-scope*, the variable is *safe-to-return* by reference. These three values form a nesting relationship from narrowest (*block-in-which-variable-is-declared*) to widest (*caller-scope*).
+
+An expression whose type is not a `ref struct` type is *safe-to-return* from the entire enclosing method. Otherwise we refer to the rules below.
```
As previous comment, what exactly is escaping?

Nigel-Ecma commented on standard/variables.md:
```diff
@@ -901,3 +901,141 @@ variable_reference
+
+For a local variable `l`:
+
+- If `v` is a `ref` variable, its *ref-safe-to-escape-scope* is taken from the *ref-safe-to-escape-scope* of its initializing expression.
```
Typo maybe, `v` here, `l` in previous line?

Nigel-Ecma commented on standard/variables.md:
```diff
@@ -901,3 +901,141 @@ variable_reference
+
+The *safe-to-escape-scope* of an expression that designates the use of a local variable is determined as follows:
+
+- A local whose type is not a `ref struct` type is *safe-to-return* from the entire enclosing method. Its *safe-to-escape-scope* is the *caller-scope*.
```
Surely a local *variable* is never *safe-to-return*, it is the contents/value of that local which is. Given we are explicitly talking about *references to variables/locations* vs. *contents/value of variables/locations" we need to be careful of the language.

BillWagner replied:
This should be value.

Nigel-Ecma reviewed (changes requested):
I haven't done a thorough review today but did find a possible typo and, to me at least, a lack of distinction between a variable/location and it’s contents. I know this stuff can be hard to comprehend for many (a student) – variables have names & values, and a name can itself be a value (aka reference, pointer, address – words which may have different semantics depending on the language). We just need to be clear *what* is escaping etc. using whatever language we decide or it will be confusing to many.

BillWagner committed:
respond to feedback

Respond to existing feedback.

BillWagner committed:
Introduce definitions

Introduce better definitions for the "lifetime" of reference variables. I avoided the term "lifetime' because of its runtime connotation. Instead, I used the "scope" where a "variable declaration" is valid.

From there, next define the safe scopes and the ref safe scopes for different variable classifications.

Finally, I shortened the terms *safe-to-escape-scope* and *ref-safe-to-escape-scope* to "*safe-scope* and *ref-safe-scope*. The text doesn't make it clear why the "escape" term is used, and it doesn't seem to add clarity.

BillWagner committed:
add examples

BillWagner commented:
@jskeet I think this one is now ready. I'd like to have async reviews before our April meeting. If so, I can clean up the remaining ref related PRs for V7.

jskeet commented:
Sounds good to me. We can discuss it in today's meeting to try to get anything sync out of the way.

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,181 @@ variable_reference
+
+### §ref-span-safety-general General
+
+A *reference variable* is a variable that refers to another variable, called the *referent*. Therefore, a *reference variable* does not store its value. Instead, a *reference variable* refers to its *referent*. Thus, the value of a reference variable is always the same as its referent variable. A *reference variable* is a local variable declared with the `ref` modifier, or an instance of a `ref s ...
```
I'm not sure what's meant by "or an instance of a `ref struct` type" here. Is an instance of a `ref struct` type always a variable? Is the point that if we call a method that returns a `ref struct`, we can treat that return value as a reference variable? An example would at least help me understand this - not sure yet whether it needs to be in the standard though!

BillWagner replied:
This is an area where the standard language gets much easier in C# 11. In C# 11, a `ref struct` can contain `ref` fields. Therefore, returning a `ref struct` by value returns the `ref` field, which is a reference.

In C# 7.n (up to C# 11), types such as `Span<T>` have *ref-like* fields. They are effectively references, but the compiler doesn't allow it yet, so it's a special case.

jskeet replied:
Hmm. It's still unclear to me how the end of this sentence works. We may be able to leave it as a fudge for now though.

BillWagner replied:
I added three examples in this section: one for a local `ref` variable. One for an instance of a `ref struct`, and one for a `return ref`. I think the `ref struct` sample should address the "fudginess".

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,181 @@ variable_reference
+
+### §ref-span-safety-general General
+
+A *reference variable* is a variable that refers to another variable, called the *referent*. Therefore, a *reference variable* does not store its value. Instead, a *reference variable* refers to its *referent*. Thus, the value of a reference variable is always the same as its referent variable. A *reference variable* is a local variable declared with the `ref` modifier, or an instance of a `ref s ...
```
Should some of these italicised terms be bold as well (as we're defining them)? @RexJaeschke to comment...

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,181 @@ variable_reference
+
+A *reference variable* is a variable that refers to another variable, called the *referent*. Therefore, a *reference variable* does not store its value. Instead, a *reference variable* refers to its *referent*. Thus, the value of a reference variable is always the same as its referent variable. A *reference variable* is a local variable declared with the `ref` modifier, or an instance of a `ref s ...
+
+A `ref struct` may include `ref struct` or ref-like fields. A ref-like field refers to the same storage as its initializing variable. Its storage size in the ref struct is the same storage size of a reference field. Unlike a reference field, a ref-like field may refer to a `struct` whose storage may be on the execution stack. Ref struct types include `Span<T>`, `ReadOnlySpan<T>`, and other types  ...
```
Ref-like in bold and italics where it's defined? (It's not clear whether the second sentence is meant to be a definition or not.)

BillWagner replied:
I intended the second sentence to be the definition. @RexJaeschke to comment on style.

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,181 @@ variable_reference
+
+A `ref struct` may include `ref struct` or ref-like fields. A ref-like field refers to the same storage as its initializing variable. Its storage size in the ref struct is the same storage size of a reference field. Unlike a reference field, a ref-like field may refer to a `struct` whose storage may be on the execution stack. Ref struct types include `Span<T>`, `ReadOnlySpan<T>`, and other types  ...
+
+A *reference return* is the expression returned by reference from a method whose return type includes the `ref` or `ref readonly` modifiers (§14.6.1). The variable expression of the *reference return* is the *referent* of the *reference return*.
```
Definitely worth an example here (eventually; doesn't need to be in this PR).

BillWagner replied:
Should I break up or move the examples starting on line 934? If so, I'll make those edits in this PR.

jskeet replied:
I think just a single example of this would be useful - just to anchor the terms "reference return" and "referent" in something concrete.

BillWagner replied:
Three examples added.

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,181 @@ variable_reference
+
+A *reference return* is the expression returned by reference from a method whose return type includes the `ref` or `ref readonly` modifiers (§14.6.1). The variable expression of the *reference return* is the *referent* of the *reference return*.
+
+All reference variables obey safety rules that ensure the valid scope of the reference variable is not greater than the valid scope of its referent.
```
Have we defined what "valid scope" means? We definitely know what a scope is, but I'm not sure whether valid scope is different.

BillWagner replied:
Re-wrote this sentence and moved to the first paragraph in the next section.

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,181 @@ variable_reference
+
+There are three valid *ref_safe_scope*s:
+
+- *block*: A variable declared in a block is valid in that block in which it is declared. This includes all local variables. A local variable declared in a method is valid in the block that defines the method. A variable declared in a block is a valid referent only if the reference variable is declared in the same block, or a nested block.
```
```suggestion
- *block*: A variable declared in a block is valid in the block in which it is declared. This includes all local variables. A local variable declared in a method is valid in the block that defines the method. A variable declared in a block is a valid referent only if the reference variable is declared in the same block, or a nested block.
```

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,181 @@ variable_reference
+There are three valid *ref_safe_scope*s:
+
+- *block*: A variable declared in a block is valid in that block in which it is declared. This includes all local variables. A local variable declared in a method is valid in the block that defines the method. A variable declared in a block is a valid referent only if the reference variable is declared in the same block, or a nested block.
+- *method*: A value parameter on a method declaration, including the implicit `this` parameter is valid in the entire method. The fields of a `struct` type are valid in the entire method. A variable with *ref_safe_scope* of *method* is a valid referent only if the reference variable is declared in the same method.
```
Does this include constructors? (What about constructor initializers?) Are indexers relevant here, given that they have parameters too?

BillWagner replied:
Yes to all these. Also, the accessors for properties. What term would be better?

jskeet replied:
Function member?

BillWagner replied:
Based on this edit, I changed the term from *method* to *function member*.

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,181 @@ variable_reference
+
+- *block*: A variable declared in a block is valid in that block in which it is declared. This includes all local variables. A local variable declared in a method is valid in the block that defines the method. A variable declared in a block is a valid referent only if the reference variable is declared in the same block, or a nested block.
+- *method*: A value parameter on a method declaration, including the implicit `this` parameter is valid in the entire method. The fields of a `struct` type are valid in the entire method. A variable with *ref_safe_scope* of *method* is a valid referent only if the reference variable is declared in the same method.
+- *caller_scope*: Member fields of a `class` type and `ref` parameters have *ref_safe_scope* of *caller_scope*. A variable with *ref_safe_scope* of *caller_scope* can be the referent of a reference return. A variable that can be referent of a *reference-return* is *safe-to-ref-return*.
```
```suggestion
- *caller_scope*: Member fields of a `class` type and `ref` parameters have *ref_safe_scope* of *caller_scope*. A variable with *ref_safe_scope* of *caller_scope* can be the referent of a reference return. A variable that can be the referent of a *reference-return* is *safe-to-ref-return*.
```

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,181 @@ variable_reference
+
+These values form a nesting relationship from narrowest (*block*) to widest (*caller-scope*).
+
+For any variable, the *safe_scope* of that variable is the scope where its value may be copied. A variable whose type is not a `ref struct` type is *safe-to-return* from the entire enclosing method. Its *safe_scope* is *caller_method*. Otherwise the following rules apply.
```
Should *caller_method* here be *caller_scope*? This is the only occurrence of *caller_method*.

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,181 @@ variable_reference
+> {
+>     private int[] arr = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
+> 
+>     public ref int M1(ref int rv)
```
I'd suggest avoid duplicating variable names. We've got `rv` in three different places (two parameters and a local variable).

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,181 @@ variable_reference
+>
+> *end example.*
+
+### §ref-span-safety-locals Local variable safe scope
```
Do we have sub-heading 4 sections? If so, I think this probably should be one - otherwise "Local variable safe scope" is a peer of "Safe scopes" when I'd expect it to be a child.

jskeet replied:
(Or possibly this just shouldn't be a new section?)

BillWagner replied:
These should all have been H4. Fixed

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,181 @@ variable_reference
+
+### §ref-span-safety-locals Local variable safe scope
+
+For a local variable `v`:
```
I was surprised to see the rule for determining a ref_safe_scope in a section about safe scopes

BillWagner replied:
Aha.  This gives me an idea:

The concept of *safe_scope* only is interesting with respect to `ref struct` types. For all other types, you can obviously pass by value and return by value. A `ref struct` is interesting ***because*** it's *safe_scope* is constrained by the *ref_safe_scope* of any member *ref-like* fields.

After I make the other edits. I'd like your thoughts on the following:

- Don't introduce *safe_scope* as a term.
- Instead, add a rule that any instance of a `ref struct` is constrained to the *ref_safe_scope* of the referent in any *ref_like_fields*.

jskeet replied:
I can't validate that that'll work, but I *do* like the idea of only having one term here. I'm finding the two confusing.

jaredpar replied:
Using one term should work just fine as long as we are consistent. Considered doing this once or twice in the ref safety docs when we implemented `ref` fields but was too much work to go back and change the original spec.

Slight tangent ... if you think about `ref` as a type vs. an annotation on a location it often makes the rules around ref safety easier to grok (at least to me). Consider that if we had `ref<int>` instead of `ref int` then having a single term becomes completely natural here as everything is just a value.

jskeet replied:
Thanks @jaredpar - that's really interesting. On my next pass through this, I'll try reading it from that perspective. @BillWagner if that turns out to be positive, *maybe* we could add a note? It's not the kind of note we'd often add, but given the complexities of this, it might be worth it here.

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,181 @@ variable_reference
+
+For a local variable `v`:
+
+- If `v` is a `ref` variable, its *ref-safe-scope* is taken from the *ref-safe-scope* of its initializing expression.
```
This section, and the ones below, use hyphens instead of underscores, which I assume is a mistake?

BillWagner replied:
Fixed. Also, swept for the term changes.

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,181 @@ variable_reference
+
+For an expression that designates the value of a formal parameter `p`, its *safe-scope* is the calling method. This applies to the `this` parameter.
+
+### §ref-span-safety-field-reference Field reference escape scope
```
Should this be "ref safe scope" rather than "reference escape scope"?

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,181 @@ variable_reference
+
+All reference variables obey safety rules that ensure the valid scope of the reference variable is not greater than the valid scope of its referent.
+
+### §ref-span-safety-escape-scopes Safe scopes
```
Perhaps call this "Safe scopes and ref-safe scopes"?

BillWagner replied:
For the next commit, I went with "Ref safe scopes".

Leaving this unresolved because of the question I have on avoiding *safe_scope* as a normative term.

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,181 @@ variable_reference
+
+These values form a nesting relationship from narrowest (*block*) to widest (*caller-scope*).
+
+For any variable, the *safe_scope* of that variable is the scope where its value may be copied. A variable whose type is not a `ref struct` type is *safe-to-return* from the entire enclosing method. Its *safe_scope* is *caller_method*. Otherwise the following rules apply.
```
I think it would be worth moving the definition of safe_scope to be nearer the definition of ref_safe_scope. Maybe.

BillWagner replied:
Leaving unresolved. See my other comment on *safe_scope*

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,181 @@ variable_reference
+
+### §ref-span-safety-limitations Limitations on reference variables
+
+- Neither a `ref` parameter, nor a `ref` local, nor a parameter or local of a `ref struct` type can be lifted into a lambda expression or local function.
```
Reference for lifting? (Is this the same lifting as in 10.6.2 and 11.4.8?)

BillWagner replied:
That's an incorrect term. It should be "captured by"

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,181 @@ variable_reference
+- A `ref struct` type may not be used as a type argument, or as an element type in a tuple type.
+- A `ref struct` type may not be the declared type of a field, except that it may be the declared type of an instance field of another `ref struct`.
+- A `ref struct` type may not be the element type of an array.
+- A value of a `ref struct` type may not be boxed:
```
We probably need to modify 10.2.9 for this. (It may be in your follow-up list already.) Various other restrictions below probably need propagating as well.

BillWagner replied:
I'd like to defer these references to these PRs:

- #44 out variables (which may just be closed now)
- #213 ref locals and returns
- #219 `in` parameters
- #601 `ref struct` types

I think the text should be in the section that defines each of those variable types. These rules may be removed at that point.

jskeet replied:
Right, that makes sense. (Although we could potentially keep these rules here as a note with references - it's handy to have them all in one place.)

jskeet replied:
Action from 2023-04-19: agree to defer.

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,181 @@ variable_reference
+For a formal parameter `p`:
+
+- If `p` is a `ref`, or `in` parameter, its *ref-safe-scope* is the *caller-scope*. It is *safe-to-return* by ref. If `p` is an `in` parameter, it can't be returned as a writable `ref` but can be returned as `ref readonly`.
+- If `p` is an `out` parameter, its *ref-safe-to-escape-scope* is the *current-method*. It isn't *safe-to-return* by ref.
```
I think this is the first time ref-safe-to-escape has been mentioned, and I see various other references to safe-to-escape-scope which hasn't been defined.

If we could make sure that every term has a definition in bold and italic, it'll be easier to make sure we never use anything that's not defined.

BillWagner replied:
Thanks! Updated all references in this section to use the newer proposed term

jskeet reviewed (commented):
A bunch of comments that I hope are useful. The whole thing bamboozles me anyway, but I *think* it's also harder to read because a name change may have been incompletely implemented. (Either that, or we need to define safe-to-escape-scope somewhere...)

I'm hoping that next time I look, it'll make more sense to me :) (I know the gist of what we're trying to achieve, but the terminology somehow just won't stick in my brain.)

BillWagner reviewed (commented):
responding to initial review

BillWagner committed:
respond to feedback.

BillWagner committed:
Apply suggestions from code review

Co-authored-by: Jon Skeet <skeet@pobox.com>

jskeet commented:
Just as an FYI: I'll look at this again tomorrow UK time, and answer questions then.

jskeet commented on standard/variables.md:
```diff
@@ -1033,36 +1033,36 @@ For a variable `c` resulting from a ref-returning method invocation, `ref e1.M(e
 A `new` expression that invokes a constructor obeys the same rules as a method invocation (§ref-span-safety-method-invocation) that is considered to return the type being constructed.
 
-In addition, for a value `c` that is the result of a `new` expression, the *safe-scope* is no wider than the smallest of the *safe-to-escape-scopes* of all arguments and operands of the object initializer expressions, recursively, if any initializer is present.
+In addition, for a value `c` that is the result of a `new` expression, the *safe_scope* is no wider than the smallest of the *safe-scopes* of all arguments and operands of the object initializer expressions, recursively, if any initializer is present.
```
```suggestion
In addition, for a value `c` that is the result of a `new` expression, the *safe_scope* is no wider than the smallest of the *safe_scope*s of all arguments and operands of the object initializer expressions, recursively, if any initializer is present.
```

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,181 @@ variable_reference
+
+There are three valid *ref_safe_scope*s:
+
+- *block*: A variable declared in a block is valid in the block in which it is declared. This includes all local variables. A local variable declared in a method is valid in the block that defines the method. A variable declared in a block is a valid referent only if the reference variable is declared in the same block, or a nested block.
```
This defines whether variables themselves are valid... which seems a bit weird. Should this be:

> A *variable_reference* to a local variable declared in a block is valid in the block in which it is declared.

Likewise for the third sentence. The last sentence looks okay already - we could rephrased all of them just by changing "is valid" to "is a valid referent" potentially.

BillWagner replied:
Updated to the suggested wording.

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,181 @@ variable_reference
+
+For an operator that yields a value, such as `e1 + e2` or `c ? e1 : e2`, the *safe_scope* of the result is the narrowest scope among the *safe_scopes* of the operands of the operator.  As a consequence, for a unary operator that yields an value, such as `+e`, the *safe_scope* of the result is the *safe_scope* of the operand.
+
+For an operator that yields a variable, such as `c ? ref e1 : ref e2`:
```
When referring to the operands here, I suspect it's only the second and third operands that are relevant, right? It doesn't matter what the safe_scope or ref_safe_scope of `c` is, does it?

jskeet replied:
(This is still a live comment.)

BillWagner replied:
I made a change, but I'm still not thrilled. Suggestions welcome.

jskeet replied:
In meeting: are the only relevant operators `?:` and `= ref`? In which case, let's just describe those directly.

jskeet commented on standard/variables.md:
```diff
@@ -901,3 +901,181 @@ variable_reference
+#### §ref-span-safety-limitations Limitations on reference variables
+
+- Neither a `ref` parameter, nor a `ref` local, nor a parameter or local of a `ref struct` type can be captured by lambda expression or local function.
+- Neither a `ref` parameter nor a parameter of a `ref struct` type may be an argument for an iterator method or an `async` method.
```
I think we typically use "shall" instead of "may", but I defer to @RexJaeschke on that.

BillWagner replied:
Defer to @RexJaeschke because this is worded as a negative, so it's mandating a compiler error. 

Rex, should these bullets all be re-written in the form "... emits a compiler error"?

jskeet replied:
Action: @BillWagner to fix both this and the line above to be unambiguous.

jskeet reviewed (commented):
I think I'm getting closer to approving :) At some point I think we'll need to just say "Good enough for now" and make any further changes later.

BillWagner committed:
address review comments

This addresses most of the comments in the most recent reviews.

The next commit will make an attempt to use a single term for *ref_safe_scope*.

BillWagner reviewed (commented):
First update to resolve many of the review comments.

BillWagner committed:
Remove *safe_scope* rules

Once I push these, I'll add notes about which rules must be added to #601

BillWagner commented on standard/variables.md:
```diff
@@ -1012,50 +1012,50 @@ A *reference variable* is a variable that refers to another variable, called the
 > *end example*
 
-A *reference return* is the expression returned by reference from a method whose return type includes the `ref` or `ref readonly` modifiers (§14.6.1). The variable expression of the *reference return* is the *referent* of the *reference return*.
+A `ref struct` may include `ref struct` or ref-like fields. A *ref-like field* refers to the same storage as its initializing expression. The initializing expression of the *ref_like field* is its *referent*. Unlike a reference field, a ref-like field may refer to a `struct` whose storage may be on the execution stack. Ref struct types include `Span<T>`, `ReadOnlySpan<T>`, and other types that in ...
```
I moved this description below the `return ref` description.

The one semantic change here says that `ref struct` types cannot be copied by value beyond the *ref_safe_scope* of its initializing expressions.

BillWagner commented on standard/variables.md:
```diff
@@ -1012,50 +1012,50 @@ A *reference variable* is a variable that refers to another variable, called the
+All reference variables obey safety rules that ensure the scope of the reference variable is not greater than the *ref_safe_scope* of its referent.
 
-For any variable, the *ref_safe_scope* of that variable is the scope where a *variable_reference* (§9.5) to that variable is valid. The *referent* of a reference variable must have a *ref_safe_scope* that is at least as wide as the scope where the reference variable is valid.
+For any variable, the *ref_safe_scope* of that variable is the scope where a *variable_reference* (§9.5) to that variable is valid. The *referent* of a reference variable must have a *ref_safe_scope* that is at least as wide as the scope of the reference variable. A `ref struct` instance may not be copied by value beyond the value *ref_safe_scope* of the initializing expression(s) of any *ref-lik ...
```
Here's the additional sentence.

jskeet replied:
I'm unclear what "of any ref-like fields" really means here. Any ref-like fields contributing to the ref struct instance? (How does the user know what's relevant here? Consider `s.AsSpan()` above - the user here doesn't know how that's implemented, or what fields are in `ReadOnlySpan<char>`. If `AsSpan` had another parameter - e.g. a `ref string x` - would that *necessarily* affect where the returned span could be copied, or would it depend on the implementation?)

(And is the word "value" in this sentence a typo? I'm not sure what that means either, if it's intentional.)

BillWagner replied:
Let's discuss some of this in our meeting. The wording is difficult until we get to C# 11, when `ref` fields are allowed. Admittedly, something like `AsSpan` is hard to deduce. Its implementation makes it more clear, in that it constructs the `Span` using a `ref char` to the first character. 

In these versions, `ref struct` types like `Span` and `ReadOnlySpan` use some attribute tricks to allow ref-like fields.

As for "value", I wanted to emphasize that returning a ref struct is returning a reference. Reworded.

BillWagner replied:
Removed from this PR, and added to the appropriate location on #601 

After thinking about Neal's points yesterday, I think the clearest way to spec this will be to have the ref safety rules in Variables. Those rules relating to the safe scope for a ref struct should be in the clause on ref structs. The rules apply, but trivially, to variables of other types. (Namely, ref structs are constrained to a safe scope. So are other variables, but for non-ref structs, the safe scope is calling method. Only ref structs have "interesting" constraints on where they can be copied.

BillWagner commented on standard/variables.md:
```diff
@@ -1110,13 +1108,6 @@ For a local variable `v`:
-- A value whose type is not a `ref struct` type is *safe_to_return* from the entire enclosing method. Its *safe_scope* is the *caller_scope*.
-- If the variable is an iteration variable of a `foreach` loop, then the variable's *safe_scope* is the same as the *safe_scope* of the `foreach` loop's expression.
-- A local of `ref struct` type which is uninitialized at the point of declaration is *safe_to_return* from the entire enclosing method. Its *safe_scope* is the *calling_method*.
-- Otherwise the variable's type is a `ref struct` type, and the variable's declaration has an initializer. The variable's *safe_scope* is the same as the *ref_safe_scope* of its initializer.
```
These two rules need to be added to #601

BillWagner commented on standard/variables.md:
```diff
@@ -1126,27 +1117,20 @@ For a formal parameter `p`:
 - Otherwise, if `p` is the `this` parameter of a struct type, its *ref_safe_scope* is the *function_member*.
 - Otherwise, the parameter is a value parameter, and it is *ref_safe_scope* is the *function_member*.
 
-For an expression that designates the value of a formal parameter `p`, its *safe_scope* is the calling method. This applies to the `this` parameter.
```
I think this just works for `ref struct` types. But, check in #601

BillWagner commented on standard/variables.md:
```diff
@@ -1126,27 +1117,20 @@ For a formal parameter `p`:
 - If `e` is of a reference type, its *ref_safe_scope* is the *caller_scope*.
 - Otherwise, if `e` is of a value type, its *ref_safe_scope* is taken from the *ref_safe_scope* of `e`.
 
-For an expression that designates a reference to a field, `e.F`, its a *safe_scope* is the same as the *safe_scope* of `e`.
```
Add this rule to #601

BillWagner commented on standard/variables.md:
```diff
@@ -1126,27 +1117,20 @@ For a formal parameter `p`:
 
 The application of a user-defined operator is treated as a method invocation.
 
-For an operator that yields a value, such as `e1 + e2` or `c ? e1 : e2`, the *safe_scope* of the result is the narrowest scope among the *safe_scopes* of the operands of the operator.  As a consequence, for a unary operator that yields an value, such as `+e`, the *safe_scope* of the result is the *safe_scope* of the operand.
```
Add this rule to #601

BillWagner commented on standard/variables.md:
```diff
@@ -1126,27 +1117,20 @@ For a formal parameter `p`:
 For an operator that yields a variable, such as `c ? ref e1 : ref e2`:
 
 - The *ref_safe_scope* of the result is the narrowest scope among the *ref_safe_scopes* of the operands of the operator.
-- The *safe_scope* of the operands must agree, and that is the *safe_scope* of the resulting variable.
```
Add this rule to #601

BillWagner commented on standard/variables.md:
```diff
@@ -1172,11 +1156,6 @@ For a variable `c` resulting from a ref-returning method invocation, `ref e1.M(e
-
-- The *caller_scope*
-- The *safe_scope* of all argument expressions (including the receiver)
-
```
This rule needs to be added to #601

BillWagner commented on standard/variables.md:
```diff
@@ -1185,20 +1164,10 @@ A value's *ref_safe_scope* is the nearest enclosing scope.
 
-#### §ref-span-safety-stackalloc stackalloc
-
-The *safe_scope* for a value `c` resulting from a `stackalloc` expression is the current method.
```
This needs to be added to either #601 or possibly #238

BillWagner commented on standard/variables.md:
```diff
@@ -1185,20 +1164,10 @@ A value's *ref_safe_scope* is the nearest enclosing scope.
 
 A `new` expression that invokes a constructor obeys the same rules as a method invocation (§ref-span-safety-method-invocation) that is considered to return the type being constructed.
 
-In addition, for a value `c` that is the result of a `new` expression, the *safe_scope* is no wider than the smallest of the *safe-scope*s of all arguments and operands of the object initializer expressions, recursively, if any initializer is present.
```
This rule needs to be added to #601

BillWagner commented on standard/variables.md:
```diff
@@ -1185,20 +1164,10 @@ A value's *ref_safe_scope* is the nearest enclosing scope.
-
-#### §ref-span-safety-default-expressions Default expressions
-
-The *safe_scope* of a `default` expression is the *caller-scope*.
```
Add this rule to #601

BillWagner commented on standard/variables.md:
```diff
@@ -1214,7 +1183,4 @@ The *safe_scope* of a `default` expression is the *caller-scope*.
 - For a ref return statement `return ref e1`, the *ref_safe_scope* of `e1` must be the *caller_scope*. In other words, `e1` must be *safe_to_return*.
-- For a return statement `return e1`, the *safe_scope* of `e1` must be the *caller_scope*. This is always true when `e1` is not a `ref struct`.
-- For an assignment `e1 = e2`, if the type of `e1` is a `ref struct` type, then the *safe_scope* of `e2` must be at least as wide a scope as the *safe_scope* of `e1`.
-- For a method invocation if there is a `ref` or `out` argument of a `ref struct` type (including the receiver), with *safe_to_escape_scope* E1, then no argument (including the receiver) may have a narrower *safe_scope* than E1.
```
These rules need to be added to #601

BillWagner reviewed (commented):
Tagging @jaredpar @gafter @cston @RikkiGibson to look at the diffs for #d8dd5d4523dc88c51ad2dc7888968bb5659ad0de

From an earlier conversation, I'm splitting the rules for *ref_safe_to_escape* (not referred to as *ref_safe_scope*, and *safe_to_escape*. 

The reason is that the *safe_to_escape* rules are only interesting in how they apply to `ref struct` types. Any other variable type can be copied (by value) anywhere.

A `ref struct` is constrained: It can only be copied within the *ref_safe_scope* of its *referent*. Moving these rules into the section on `ref structs` makes that distinction more clear.

Nigel-Ecma commented on standard/variables.md:
```diff
@@ -832,7 +832,7 @@ For an expression *expr* of the form:
 
 - The definite-assignment state of *v* before *expr_cond* is the same as the state of *v* before *expr*.
-- The definite-assignment state of *v* before *expr_true* is definitely assigned if the state of *v* after *expr_cond* is definitely assigned or “definitely assigned after true expression”.
+- The definite-assignment state of *v* before *expr_true* is definitely assigned if the state of *v* after *expr_cond* is definitely assigned or “definitely assigned after true expression”1057.
```
I'm guessing this 1057 is an accident.

```suggestion
- The definite-assignment state of *v* before *expr_true* is definitely assigned if the state of *v* after *expr_cond* is definitely assigned or “definitely assigned after true expression”.
```

Nigel-Ecma reviewed (changes requested):
Just a single comment but git decided it needed to be a review…

BillWagner committed:
Update standard/variables.md

Co-authored-by: Nigel-Ecma <perryresearch@zoot.net.nz>

BillWagner commented on standard/variables.md:
```diff
@@ -987,3 +987,200 @@ variable_reference
+  - No instance method of a `ref struct` type may be captured by method conversion to a delegate type.
+- For a ref reassignment `ref e1 = ref e2`, the *ref_safe_scope* of `e2` must be at least as wide a scope as the *ref_safe_scope* of `e1`.
+- For a ref return statement `return ref e1`, the *ref_safe_scope* of `e1` must be the *caller_scope*. In other words, `e1` must be *safe_to_return*.
+- A local function or anonymous function may not refer to a local or parameter of `ref struct` type declared in an enclosing scope.
```
Let's put this one back here:

```suggestion
- For a method invocation if there is a `ref` or `out` argument of a `ref struct` type (including the receiver), with *ref_safe_scope* E1, then no argument (including the receiver) may have a narrower *ref_safe_scope* than E1.
- A local function or anonymous function may not refer to a local or parameter of `ref struct` type declared in an enclosing scope.
```

BillWagner committed:
Update standard/variables.md

jskeet commented on standard/variables.md:
```diff
@@ -987,3 +987,201 @@ variable_reference
+
+A *reference return* is the expression returned by reference from a method whose return type includes the `ref` or `ref readonly` modifiers (§15.6.1). The variable expression of the *reference return* is the *referent* of the *reference return*.
+
+> *Example:* The following example demonstrates a =*reference return* whose *referent* is an element of an array field:
```
Nit: remove `=` before *reference return*.

jskeet commented on standard/variables.md:
```diff
@@ -987,3 +987,201 @@ variable_reference
+>
+> *end example*
+
+A `ref struct` may include `ref struct` or ref-like fields. A *ref-like field* refers to the same storage as its initializing expression. The initializing expression of the *ref_like field* is its *referent*. Unlike a reference field, a ref-like field may refer to a `struct` whose storage may be on the execution stack. Ref struct types include `Span<T>`, `ReadOnlySpan<T>`, and other types that in ...
```
ref_like => ref-like

jskeet commented on standard/variables.md:
```diff
@@ -987,3 +987,201 @@ variable_reference
+
+A `ref struct` may include `ref struct` or ref-like fields. A *ref-like field* refers to the same storage as its initializing expression. The initializing expression of the *ref_like field* is its *referent*. Unlike a reference field, a ref-like field may refer to a `struct` whose storage may be on the execution stack. Ref struct types include `Span<T>`, `ReadOnlySpan<T>`, and other types that in ...
+
+> *Example:* The following example demonstrates a `ReadOnlySpan<T>` *reference variable* whose *referent* is the characters in a string. `ReadOnlySpan<T>` is a `ref struct`. One of its internal fields is a *ref-like field* that refers to the character array in the `string` referent. Therefore, `chars` is a *reference variable*:
```
Earlier we have: "A *reference variable* is a local variable declared with the `ref` modifier."

... but `chars` is apparently a reference variable, despite not being declared with the ref modifier. Perhaps the earlier sentence needs to include "or a variable whose type is a ref struct"?

BillWagner replied:
I reworded the text of the example. In this clause, I want to introduce that `ref struct` types contain reference variables. Therefore, they must not be copied outside of the ref safe scope of those reference variables.

jskeet commented on standard/variables.md:
```diff
@@ -987,3 +987,201 @@ variable_reference
+
+All reference variables obey safety rules that ensure the scope of the reference variable is not greater than the *ref_safe_scope* of its referent.
+
+For any variable, the *ref_safe_scope* of that variable is the scope where a *variable_reference* (§9.5) to that variable is valid. The *referent* of a reference variable must have a *ref_safe_scope* that is at least as wide as the scope of the reference variable. A `ref struct` instance may not be copied by value beyond the value *ref_safe_scope* of the initializing expression(s) of any *ref-lik ...
```
I'm not sure why we've got a mixture of dashes and underscores here. I *think* I'd expect dashes in descriptive names, but underscores in grammar elements. Would welcome @Nigel-Ecma's view here.

jskeet commented on standard/variables.md:
```diff
@@ -987,3 +987,201 @@ variable_reference
+
+There are three valid *ref_safe_scope*s:
+
+- *block*: A *variable_reference* to a local variable declared in a block is valid in the block in which it is declared. The *ref_safe_scope* of a local variable is *block*. A local variable declared in a method has a *ref_safe_scope* of the *block* that defines the method. A variable declared in a block is a valid referent only if the reference variable is declared in the same block, or a nested ...
```
> A *variable_reference* to a local variable declared in a block is valid in the block in which it is declared.

Does it have to be definitely assigned? Presumably it's not valid prior to the declaration itself, i.e. I wouldn't *expect*:

```csharp
ref int x = ref y;
int y;
```

... to compile.

BillWagner replied:
Great catch. That doesn't compile. Updated the text.

jskeet commented on standard/variables.md:
```diff
@@ -987,3 +987,201 @@ variable_reference
+
+There are three valid *ref_safe_scope*s:
+
+- *block*: A *variable_reference* to a local variable declared in a block is valid in the block in which it is declared. The *ref_safe_scope* of a local variable is *block*. A local variable declared in a method has a *ref_safe_scope* of the *block* that defines the method. A variable declared in a block is a valid referent only if the reference variable is declared in the same block, or a nested ...
```
> A local variable declared in a method has a *ref_safe_scope* of the *block* that defines the method.

What if the method doesn't have a block declaration, i.e. it's expression-bodied. I don't know whether an out parameter would count as "a local variable declared in a method"... this may not be important if you can't declare a reference variable other than in a block.

jskeet commented on standard/variables.md:
```diff
@@ -987,3 +987,201 @@ variable_reference
+
+There are three valid *ref_safe_scope*s:
+
+- *block*: A *variable_reference* to a local variable declared in a block is valid in the block in which it is declared. The *ref_safe_scope* of a local variable is *block*. A local variable declared in a method has a *ref_safe_scope* of the *block* that defines the method. A variable declared in a block is a valid referent only if the reference variable is declared in the same block, or a nested ...
```
> A local variable declared in a method has a *ref_safe_scope* of the *block* that defines the method.

So something like this is presumably okay:

```csharp
public void M(string x)
{
    ReadOnlySpan<char> span1;
    if (x is not null)
    {
        ReadOnlySpan<char> span2 = x.AsSpan();
        span1 = span2;
    }
}
```

... because the ref_safe_scope of span2 is the whole of the method's block, not just the block of the if statement. (I haven't tested that - my machine is chugging along doing a large build at the moment.)

BillWagner replied:
Yes, that's valid.  See https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQGAzAATakDCpA3oSeSqQLIAUGOADKQgJSF0CpYaQBKMAIYATAPIA7ADYBPAMoAHCXIA8YABYSoAPlIBnDXJwBuQiNIBLAGalWCeydJyIAFw8BXBQr8QiKCtrbi0vLK6po6+kam5likALw8AHQAgiYxcqy81sFhiZo4qSVyWIW2AL6EdQQ1QA

jskeet commented on standard/variables.md:
```diff
@@ -987,3 +987,201 @@ variable_reference
+
+- *block*: A *variable_reference* to a local variable declared in a block is valid in the block in which it is declared. The *ref_safe_scope* of a local variable is *block*. A local variable declared in a method has a *ref_safe_scope* of the *block* that defines the method. A variable declared in a block is a valid referent only if the reference variable is declared in the same block, or a nested ...
+- *function_member*: A *variable_reference* to a value parameter on a function member declaration, including the implicit `this` parameter is valid in the entire function member. The *ref_safe_scope* of the fields of a `struct` type is *function_member*. A variable with *ref_safe_scope* of *function_member* is a valid referent only if the reference variable is declared in the same function member ...
+- *caller_scope*: Member fields of a `class` type and `ref` parameters have *ref_safe_scope* of *caller_scope*. A variable with *ref_safe_scope* of *caller_scope* can be the referent of a reference return. A variable that can be the referent of a *reference-return* is *safe-to-ref-return*.
```
Do we need the last sentence here? This appears to be the only use of safe-to-ref-return.

jskeet replied:
Ah - that's because we later have *safe_to_return* by ref. So presumably this should be *safe_to_return* by ref? (I think we need a definition for this somewhere, and ideally making it a single italicised phrase - I personally prefer safe-to-ref-return, so maybe we should change everything else instead of this.)

BillWagner replied:
~~I am confused by these two comments. Let's discuss at our meeting.~~

I figured it out from other comments. Replaced *safe-to-return* by ref with *safe-to-ref-return* throughout this PR.

jskeet commented on standard/variables.md:
```diff
@@ -987,3 +987,201 @@ variable_reference
+
+These values form a nesting relationship from narrowest (*block*) to widest (*caller_scope*).
+
+> *Example*: The following code shows examples of the different *ref_safe_scope* values. The declarations show values of *ref_safe_scope* for a referent to be the initializing expression for a `ref` variable. The examples show the values of *ref_safe_scope* for a reference return:
```
Do we need the word "values" in this sentence? A ref_safe_scope feels a little less tangible than a normal value. Just *ref_safe_scope*s would work for me. (Likewise for the second sentence.)

jskeet commented on standard/variables.md:
```diff
@@ -987,3 +987,201 @@ variable_reference
+
+For a local variable `v`:
+
+- If `v` is a `ref` variable, its *ref_safe_scope* is taken from the *ref_safe_scope* of its initializing expression.
```
We've got "`ref` variable" here where earlier we've referred to "reference variable". If we mean "reference variable" here, please update it - importantly, `ReadOnlySpan<char> span = s.AsSpan();` is apparently a reference variable, but really doesn't look like a `ref` variable.

jskeet replied:
And could "is taken from" be "is the same as"? (That feels more precise to me.)

jskeet commented on standard/variables.md:
```diff
@@ -987,3 +987,201 @@ variable_reference
+- If `p` is a `ref`, or `in` parameter, its *ref_safe_scope* is the *caller_scope*. It is *safe_to_return* by ref. If `p` is an `in` parameter, it can't be returned as a writable `ref` but can be returned as `ref readonly`.
+- If `p` is an `out` parameter, its *ref_safe_scope* is the *function_member*. It isn't *safe_to_return* by ref.
+- Otherwise, if `p` is the `this` parameter of a struct type, its *ref_safe_scope* is the *function_member*.
+- Otherwise, the parameter is a value parameter, and it is *ref_safe_scope* is the *function_member*.
```
it is => its... and specify that it isn't safe to return by ref?

jskeet commented on standard/variables.md:
```diff
@@ -987,3 +987,201 @@ variable_reference
+
+- If `p` is a `ref`, or `in` parameter, its *ref_safe_scope* is the *caller_scope*. It is *safe_to_return* by ref. If `p` is an `in` parameter, it can't be returned as a writable `ref` but can be returned as `ref readonly`.
+- If `p` is an `out` parameter, its *ref_safe_scope* is the *function_member*. It isn't *safe_to_return* by ref.
+- Otherwise, if `p` is the `this` parameter of a struct type, its *ref_safe_scope* is the *function_member*.
```
And is it safe to ref return?

BillWagner replied:
Yes. ~~Leaving unresolved to discuss which term is preferred in our meeting.~~ Updated to *safe-to-ref-return*

jskeet commented on standard/variables.md:
```diff
@@ -987,3 +987,201 @@ variable_reference
+For a variable designating a reference to a field, `e.F`:
+
+- If `e` is of a reference type, its *ref_safe_scope* is the *caller_scope*.
+- Otherwise, if `e` is of a value type, its *ref_safe_scope* is taken from the *ref_safe_scope* of `e`.
```
Again, I'd like a bit more clarity than "taken from" if possible.

jskeet reviewed (commented):
Much clearer than before, thanks. I've raised various comments, but I think my overall feeling is:

- We need another pass at consistency of terms and dash-separation vs underscore_separation
- It feels like we're missing a *definition* of "safe to ref return". It's possible I just missed it though!

BillWagner committed:
Respond to review comments.

jskeet commented on standard/variables.md:
```diff
@@ -987,3 +987,201 @@ variable_reference
+
+#### §ref-span-safety-operators Operators
+
+The application of a user-defined operator is treated as a method invocation.
```
In meeting: can we remove this line entirely? (User-defined operators can't return by ref or have ref parameters.)

jskeet reviewed (commented):
@gafter will have a look at this, and there are a few actions defined in comments, but otherwise, we're ready to merge.

BillWagner committed:
Feedback from April meeting

This include comments, and fixing the formatting after consulting with Rex.

BillWagner committed:
remove ref struct descriptions

These belong in the PR for ref structs, and in the section on ref structs.

RexJaeschke commented:
@ BillWagner. Per your request, I just looked over this PR and saw a potential problem in the following:

> There are three valid ref-safe-scopes:
>
> - ***block***: A *variable_reference* to a local variable declared in a block is valid from its declaration to the end of the block in which it is declared. The ref-safe-scope of a local variable is block. A local variable declared in a method has a ref-safe-scope of the block that defines the method. A variable declared in a block is a valid referent only if the reference variable is declared in the same block after the referent, or a nested block.
> - ***function-member***: A *variable_reference* to a value parameter on a function member declaration, including the implicit `this` parameter is valid in the entire function member. The ref-safe-scope of the fields of a `struct` type is function-member. A variable with ref-safe-scope of function-member is a valid referent only if the reference variable is declared in the same function member.
> - ***caller-scope***: Member fields of a `class` type and `ref` parameters have ref-safe-scope of caller-scope. A variable with ref-safe-scope of caller-scope can be the referent of a reference return. A variable that can be the referent of a reference-return is ***safe-to-ref-return***.

I'm concerned about defining a new term called ***block***, as we already have a grammar rule by that name. My first thought was to rename your 3 new terms ***ref-safe-block-scope***, ***ref-safe-function-member-scope***, and ***ref-safe-caller-scope***. However, I realize they are rather long, and perhaps a little unwieldy. But it does directly relate them.

In the definition of ***block***,  "The ref-safe-scope of a local variable is block.", if "block" is intended to be a reference to a grammar rule, it should be italicized.  Likewise for "function-member" in "The ref-safe-scope of the fields of a `struct` type is function-member." in the ***function-member*** definition ***EXCEPT*** that grammar rules use "_" rather than "-", so I'm not sure what this (and possibly "block") are referring to.

BillWagner commented:
> I'm concerned about defining a new term called _**block**_, as we already have a grammar rule by that name. My first thought was to rename your 3 new terms _**ref-safe-block-scope**_, _**ref-safe-function-member-scope**_, and _**ref-safe-caller-scope**_. However, I realize they are rather long, and perhaps a little unwieldy. But it does directly relate them.

I'll propose that we should use these terms. Yes, they are unwieldy, but so are the concepts. These terms directly relate to the concepts, as you say.

I've been looking for terms that were both succinct and descriptive, and I've always come up short. I think descriptive wins.

I'd like others thoughts before I make those edits. Tagging @gafter @jskeet @MadsTorgersen @Nigel-Ecma

Nigel-Ecma commented:
> I'll propose that we should use these terms. Yes, they are unwieldy, but so are the concepts. These terms directly relate to the concepts, as you say.
> 
> 
> 
> I've been looking for terms that were both succinct and descriptive, and I've always come up short. I think descriptive wins.
> 
> 
> 
> I'd like others thoughts before I make those edits. Tagging @gafter @jskeet @MadsTorgersen @Nigel-Ecma 

Agree, go wth the longer descriptive terms

jskeet commented on standard/variables.md:
```diff
@@ -987,3 +987,168 @@ variable_reference
+There are three valid ref-safe-scopes:
+
+- ***block***: A *variable_reference* to a local variable declared in a block is valid from its declaration to the end of the block in which it is declared. The ref-safe-scope of a local variable is block. A local variable declared in a method has a ref-safe-scope of the block that defines the method. A variable declared in a block is a valid referent only if the reference variable is declared in ...
+- ***function-member***: A *variable_reference* to a value parameter on a function member declaration, including the implicit `this` parameter is valid in the entire function member. The ref-safe-scope of the fields of a `struct` type is function-member. A variable with ref-safe-scope of function-member is a valid referent only if the reference variable is declared in the same function member.
```
```suggestion
- ***function-member***: A *variable_reference* to a value parameter on a function member declaration, including the implicit `this` parameter, is valid in the entire function member. The ref-safe-scope of the fields of a `struct` type is function-member. A variable with ref-safe-scope of function-member is a valid referent only if the reference variable is declared in the same function member.
```

(Just a comma after "`this` parameter".)

jskeet commented on standard/variables.md:
```diff
@@ -987,3 +987,168 @@ variable_reference
+- ***function-member***: A *variable_reference* to a value parameter on a function member declaration, including the implicit `this` parameter is valid in the entire function member. The ref-safe-scope of the fields of a `struct` type is function-member. A variable with ref-safe-scope of function-member is a valid referent only if the reference variable is declared in the same function member.
+- ***caller-scope***: Member fields of a `class` type and `ref` parameters have ref-safe-scope of caller-scope. A variable with ref-safe-scope of caller-scope can be the referent of a reference return. A variable that can be the referent of a reference-return is ***safe-to-ref-return***.
+
+These values form a nesting relationship from narrowest (*block*) to widest (caller-scope).
```
```suggestion
These values form a nesting relationship from narrowest (block) to widest (caller-scope).
```

jskeet commented on standard/variables.md:
````diff
@@ -987,3 +987,168 @@ variable_reference
+> *Example*: The following code shows examples of the different ref-safe-scopes. The declarations show the ref-safe-scope for a referent to be the initializing expression for a `ref` variable. The examples show the ref-safe-scope for a reference return:
+>
+> ```csharp
+> public class C
````
Possibly add the ref-safe-scope of `r1`, `v1`, `arr` and `arr[v2]`?

jskeet commented on standard/variables.md:
```diff
@@ -987,3 +987,168 @@ variable_reference
+
+- The caller-scope.
+- The ref-safe-scope of all `ref` and `out` argument expressions (excluding the receiver).
+- For each `in` parameter of the method, if there is a corresponding expression that is an variable, its ref-safe-scope, otherwise the nearest enclosing scope
```
```suggestion
- For each `in` parameter of the method, if there is a corresponding expression that is a variable, its ref-safe-scope, otherwise the nearest enclosing scope
```

jskeet commented on standard/variables.md:
```diff
@@ -987,3 +987,168 @@ variable_reference
+- For each `in` parameter of the method, if there is a corresponding expression that is an variable, its ref-safe-scope, otherwise the nearest enclosing scope
+- The scope of all argument expressions (including the receiver).
+
+> *Example*: the last bullet is necessary to handle code such as
```
It would be good to improve this example over time - I don't know what "handling" these situations really means.

BillWagner replied:
I replaced it with a more obvious example.

jskeet reviewed (approved):
Minor comments, but I think I'm still fine with it. There may well be nuances that I've missed - the whole thing is quite confusing, but that's just the nature of it.

jskeet reviewed (approved):
Minor comments, but I think I'm still fine with it. There may well be nuances that I've missed - the whole thing is quite confusing, but that's just the nature of it.

jskeet reviewed (approved):
Minor comments, but I think I'm still fine with it. There may well be nuances that I've missed - the whole thing is quite confusing, but that's just the nature of it.

jskeet reviewed (approved):
Minor comments, but I think I'm still fine with it. There may well be nuances that I've missed - the whole thing is quite confusing, but that's just the nature of it.

BillWagner committed:
Apply suggestions from code review

Co-authored-by: Jon Skeet <skeet@pobox.com>

BillWagner committed:
respond to feedback.

BillWagner committed:
Merge branch 'ref-span-safety' of https://github.com/BillWagner/csharpstandard into ref-span-safety

Nigel-Ecma commented on standard/variables.md:
```diff
@@ -987,3 +987,175 @@ variable_reference
+
+### §ref-span-safety-general General
+
+A ***reference variable*** is a variable that refers to another variable, called the ***referent***. Therefore, a reference variable does not store its value. Instead, a reference variable refers to its referent. Thus, the value of a reference variable is always the same as its referent variable. A reference variable is a local variable declared with the `ref` modifier.
```
This is a bit torturous, *"a reference variable does not store its value"* – it does, what it doesn't store is the *referent's value*. I understand what is being attempted – to not talk about the *value* stored in the ref variable as much as possible, except we have to as we can reassign that value so the variable refers to a different referent…

There is a comment below by @jaredpar, its hidden now so here's an extract: 

> Slight tangent ... if you think about ref as a type vs. an annotation on a location it often makes the rules around ref safety easier to grok (at least to me). Consider that if we had ref<int> instead of ref int then having a single term becomes completely natural here as everything is just a value.

What we're being fuzzy over is that it is a type, a reference variable holds a value of type "*reference to a variable*" – echoes of Algol-68 `ref ref int x := ref int y` ;-), but that languages attempt to make all this clear confused a *lot* of people…

Anyway I'll attempt a bit of wordsmithing here but really we need to consider how we're futzing this in general:

```suggestion
A ***reference variable*** is a variable that refers to another variable, called the ***referent***. A reference variable does not store the value of its referent. When a reference variable is used where a value is required its referent's value is returned; similarly when a reference variable is the target of an assignment it is the referent which is assigned to. The variable to which a reference variable refers, i.e. its referent, can be changed using a ref assignment (`= ref`) [insert x-ref]. A reference variable is a local variable declared with the `ref` modifier.
```
I can't say I'm totally happy with that, maybe someone else can improve it, but I think its a bit clearer.

Nigel-Ecma reviewed (changes requested):
I'm marking this a "request changes" but I'm only suggesting the change and others might manage better wordsmithing.

BillWagner committed:
Apply suggestions from code review

Co-authored-by: Nigel-Ecma <perryresearch@zoot.net.nz>

BillWagner committed:
remove missing xref

This needs to be added to #213 once this PR is merged.

MadsTorgersen commented on standard/variables.md:
```diff
@@ -987,3 +987,175 @@ variable_reference
+
+The conditional operator (§12.18), `c ? ref e1 : ref e2`, and reference assignment operator, `= ref e` (§12.21.1) have reference variables as operands and yield a reference variable. For those operators, the ref-safe-scope of the result is the narrowest scope among the ref-safe-scopes of all `ref` operands.
+
+#### §ref-span-safety-method-invocation Method invocation
```
Does this apply to other function member invocations as well, such as indexer invocations?

BillWagner replied:
Resolved at the meeting: generalize to all function invocations (local functions, property and indexer accessors, and delegates).

BillWagner commented on standard/variables.md:
```diff
@@ -987,3 +987,175 @@ variable_reference
+>
+> *end example*
+
+### §ref-span-safety-escape-scopes Ref safe scopes
```
These should not be called "scopes". 

Resolved at the meeting: Refer to these as the safe "context".  Add a *note* to explain that the compiler is inferring the valid lifetime of the variables based on the program text. The compiler determines that the lifetime would extend at least as wide as the program text of the safe context.

BillWagner commented on standard/variables.md:
```diff
@@ -987,3 +987,175 @@ variable_reference
+
+There are three valid ref-safe-scopes:
+
+- ***block***: A *variable_reference* to a local variable declared in a block is valid from its declaration to the end of the block in which it is declared. The ref-safe-scope of a local variable is block. A local variable declared in a method has a ref-safe-scope of the block that defines the method. A variable declared in a block is a valid referent only if the reference variable is declared in ...
```
Add an explanation that blocks are nested, the same way that scopes are nested. There is a different context for each block.

BillWagner replied:
addressed in next commit.

BillWagner commented on standard/variables.md:
```diff
@@ -987,3 +987,175 @@ variable_reference
+- ***function-member***: A *variable_reference* to a value parameter on a function member declaration, including the implicit `this` parameter, is valid in the entire function member. The ref-safe-scope of the fields of a `struct` type is function-member. A variable with ref-safe-scope of function-member is a valid referent only if the reference variable is declared in the same function member.
+- ***caller-scope***: Member fields of a `class` type and `ref` parameters have ref-safe-scope of caller-scope. A variable with ref-safe-scope of caller-scope can be the referent of a reference return. A variable that can be the referent of a reference-return is ***safe-to-ref-return***.
+
+These values form a nesting relationship from narrowest (block) to widest (caller-scope).
```
expand to included nested blocks.

BillWagner committed:
Updates from 5/17 meeting.

Updates from the 5/17 meeting.

BillWagner committed:
fix build warning
